bootstraps the generator plugin and compiles the plugin with its own …

menduz opened the pull request:
This PR chronologically
1. fine tunes the Go code generator and fixes some issues with the `protobuf.zig` library
2. creates a Zig native plugin for `protoc`
3. fine tunes the plugin's output until the plugin can compile itself with the output of the previous run
4. removes the go dependency
5. adds and fixes tests

Original PR and history at https://github.com/menduz/zig-protobuf/pull/1

Pending work:
- Idiomatic maps: use hash maps instead of list of key-value entries
- Default values: enable initialization of messages inheriting the default values defined in the .proto file

Arwalk commented:
Starting the review now. Might take some time considering how much work you did :)

Arwalk reviewed (changes requested):
I'd like to make one last pass on the tests, but i can leave this first review for starters. There are some minor changes suggested.

It's a lot of work and i like it for many reasons, but i have a little something that bothers me right now on the subject of allowing required fields.

Right now the implementation is supposed to only handle proto3 for the sake of simplicity, and proto3 only has optional fields.

I guess the few required fields that are around, and the code that handle it, are related to your use case. Would it be possible to:
- make the code generator output optional fields for these definitions, as the wire protocol is entirely retrocompatible
- remove the existing code that handles required fields for now

This would still make it possible to use with required field (at the expense of the strictness associated with it) and will leave us the time to handle properly proto2 in a later phase.

What do you think of this?

Once again, i can not thank you enough for all this work, and do not want to sound ungrateful. I hope you won't take my comments badly.

menduz commented on tests/fixtures/unittest.proto.bin:
deleted

menduz commented:
Many thanks for your review. I've pushed most of the suggestions and I'd like to discuss about the optional/required-ness of the fields. I also mistakenly pushed a new set of changes to the main branch. Including `oneof` support and partial support for maps using lists of entries. Also I cleaned up the code and added mixins to reduce the emitted code `(pub usingnamespace protobuf.MessageMixins(@This());`). I've also added a good amount of tests including unitary tests, and some "blackbox" tests with big encoded files.

### Optional/Required fields

In regards to optional fields, as of now, all fields are being emitted as optional unless `required` is explicitly specified in the .proto file. The wire protocol doesn't break if any of those required are omitted. The for knowing the value are relatively straight forward:
- In initialization of a message (zig), pick the default value of the field type (false, 0, null, "", enum=0). Unless a default value is specified in the .proto, in such case, pick that value.
- In "reading" a message, only change a result.value if it appears in the encoded message. Otherwise use the default value. Similar to `vat t: T = undefined;` in zig.

From my experience, the use of some required fields was not a deal breaker, but the protocol enables us to swap a `?u32` by a `u32` and I can see the benefit on that. Would you elaborate on your PoV?

## Oneof

I revamped your original code for unions and completed the encoding/decoding part and added extensive tests. Please review that new part.

## Maps

For simplicity, I've leveraged the default behavior for protoc and enabled maps via list of autogenerated entries. This is something I'd like to work a bit more but to close up some work I'd like to defer it for the next PR.

## Open questions about allocations and code speed

> Disclamer: you may notice silly things in my zig code, this is the project I'm using to learn most of the language. I do have extensive experience in other languages including C++ that I can certainly leverage right now. 

I thought about enabling all strings and submessages as pointers. Because recursive messages are currently not possible.

```diff
 const Message = struct {
-  recursive: ?Message
+  recursive: ?*Message
 }
```

But that adds a considerable overhead for the daily use, I understand that messages are currently stored entirely on the stack and are passed to functions as reference or copied, but that's out of our control. Pointers may fix this to a certain level, also it would enable more precise control of \*WHERE* in the memory we store each message. 

I was tempted to introduce it like the ArrayList does, something like the following snippet, but dozens of unanswered questions arised. Most of them are unknown unknowns, since my experience with zig is limited. I'd love to have some equivalent concepts to "std::move" to anchor some concepts.

```zig
fn Messagee(comptime allocator: Allocator) {
  return struct {
    ...
  };
} 
```

And taking some learnings from the protobuf C++ library, I think at this point we should have a special (arena) allocator for all messages. Because if we keep the recursive `deinit` semantics, we may be freeing some parts of the memory that other parts of the application are still considering alive. This exact same point applies to strings. If we swap `[]const u8` by a pointer, how do we decide over the lifecycle of that string?

This is a new talking point that I think deserves our attention.

menduz commented on src/protobuf.zig:
```diff
@@ -363,17 +430,23 @@ fn internal_pb_encode(pb: *ArrayList(u8), data: anytype) !void {
-                    try append(pb, @field(data_type._desc_table, field.name), @TypeOf(@field(data, field.name)), @field(data, field.name));
+                    try append(pb, @field(data_type._desc_table, field.name), @field(data, field.name));
+                },
+                .Varint, .FixedInt => if (@as(u64, @field(data, field.name)) != 0) {
+                    try append(pb, @field(data_type._desc_table, field.name), @field(data, field.name));
                 },
```
This is slightly off the spec (right now). The expected behavior is something like

```zig
// In this case, we are assuming that the default value of each
// field is 0 to omit it on the encoded message.
if (data.field != null and data.field != meta(data.field).default_value) {
  write_field(data.field)
}
```



In the totality of my changes, I assume required fields are unexistent. and in some cases, it may only be a hint for the typesystem, a help to the user to prevent many `.?` everywhere.

menduz commented on tests/fixtures/vector_tile.bin:
mapbox tiles, it is a format I used many years ago to condense geographical features in a super compressed format. I copied the example from another protobuf library including its license.

the conformance test decodes and re-encodes the file asserting equal results

menduz commented on tests/unittest.zig:
wip, will remove!

Arwalk commented on src/protobuf.zig:
```diff
@@ -723,81 +687,96 @@ fn decode_packed_list(slice: []const u8, comptime list_type: ListType, comptime
         },
-        .OneOf => |_| {
-            @compileError("Can not decode OneOf fields yet");
+        .OneOf => |one_of| {
+            // the following code:
+            // 1. creates a compile time for iterating over all `one_of._union_desc` fields
+            // 2. when a match is found, it creates the union value in the `field.name` property of the struct `result`. breaks the for at that point
+            const desc_union = one_of._union_desc;
+            inline for (@typeInfo(one_of).Union.fields) |union_field| {
+                const v = @field(desc_union, union_field.name);
+                if (is_tag_known(v, extracted_data)) {
+                    var value = try decode_value(union_field.type, v.ftype, extracted_data, allocator);
+                    @field(result, field.name) = @unionInit(one_of, union_field.name, value);
+                }
+            }
```
Thank you for finding a solution for this!  I remember trying decoding oneOfs at some point but i'm not sure 
`@unionInit` existed yet. Very happy about this.

Arwalk commented:
I took a look at what has been done since. Always as happy about your contributions. I'll have some time tomorrow afternoon to take a look in more details and answer our current subjects. It's a pleasure to work with you.

menduz commented:
Thanks for the detailed answer. I think we are good to merge the PR now, and based on your comment, we should.

However, I think the library leaks memory on strings. For data-section static strings it won't. But if we are using any means to allocate arrays of bytes in runtime I think it will.

Probably this should be handled right after this PR, along with the maps.

Maybe we can provide the allocator as a parameter to the deinit function, enabling us to
1) swap ArrayList for ArrayListUnmanaged
2) clean up memory on submessages
3) clear strings if those are not static. I assume there is some heuristic we can leverage to know if a pointer is static memory or dynamic. To prevent `allocator.free(const)`.

The biggest drawback may be that in order to set string values, we may need to have a `ManagedString` and `ManagedBytes` structs. Those would have copy semantics, in effect it would be something like this:

```diff
 const Message = struct {
-    m: ?[]const u8
+    m: ?ManagedString
     ...
 }
 ...
 /// allocates memory for a string and copies the provided slice
 fn managedStringFromSlice(str: []const u8, allocator: Allocator) !ManagedString
 ...
 var m = proto.Message.init(allocator);
-m.str = "my string";
+m.str = managedStringFromSlice("my string", allocator);
```

Arwalk commented:
I agree with everything in your last comment. I'll take it into account when i'll do some cleanup on the issue side. I was coming to the conclusion too that we needed something a bit specific for strings.
